torch_xla_py/xla_dist.py

```diff
@@ -402,3 +408,249 @@ def get_cluster(self):
+
+  def _prepare_scripts(self, cmd):
+    cmd_str = ' '.join(cmd)
+    for var in self.DIST_ENV_VARS:
+      if var in cmd_str:
+        raise ValueError((
+            '{} should not be in the training command provided as they'
+            ' will interfere with the values set for distributed'
+            ' training'.format(var)))
+
+    worker_script_map = {}
+    for i in range(len(self._cluster._client_workers)):
+      script_path = self.SCRIPT_PATH_TMPL.format(pid=os.getpid(), worker=i)
+
+      export_env_cmd = self._env_vars_cmd(i)
+      # Setup environment for non-interactive non-login shell over ssh
+      export_env_cmd.append('. /etc/profile;')
+      worker_cmd = cmd
+      if self.is_docker:
+        docker_vars_cmd = self._docker_vars_cmd(cmd)
+        docker_idx = cmd.index('docker')
+        run_idx = cmd.index('run')
+        assert run_idx == docker_idx+1
+        worker_cmd = cmd[:run_idx+1] + docker_vars_cmd + cmd[run_idx+1:]
+
+      script_cmd = export_env_cmd + worker_cmd
+      os.makedirs(os.path.dirname(script_path), exist_ok=True)
+      with open(script_path, 'w') as f:
+        script = ' '.join(script_cmd)
+        f.write(script)
+      subprocess.call(['chmod', '+x', script_path])
+      worker_script_map[self._cluster._client_workers[i]] = script_path
+
+    return worker_script_map
+
+  def _scp_scripts(self, script_map):
+
+    def _gcloud_scp(script_path, client_worker):
+      scp_cmd = [
+          'gcloud', '-q', 'compute', 'scp', '--internal-ip',
+          '--zone={}'.format(client_worker._zone),
+          script_path, 'pytorchtpudistrunner@{}:~/{}'.format(
+              client_worker._hostname, os.path.basename(script_path)),
+      ]
+      proc = subprocess.Popen(scp_cmd, stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE)
+      self._stream_logs(proc, client_worker)
+
+    threads = []
+    for i, client_worker in enumerate(script_map):
+      if i == 0:
+        # ssh keygen single time
+        _gcloud_scp(script_map[client_worker], client_worker)
+        continue
+      thread = threading.Thread(
+          target=_gcloud_scp,
+          args=(script_map[client_worker], client_worker,))
+      thread.daemon = True
+      thread.start()
+      threads.append(thread)
+
+    for thread in threads:
+      thread.join()
+
+  def _cleanup(self, script_path, client_worker):
+    cleanup_cmd = ['rm', '~/{}'.format(os.path.basename(script_path))]
+    if self.is_docker:
+      cleanup_cmd.extend(['&&', 'docker', 'rm', '-f', self.container_name])
+    cleanup_cmd.extend(['&&', 'pkill', '-u', 'pytorchtpudistrunner'])
+    docker_rm_cmd = [
+        'gcloud', '-q', 'compute', 'ssh', '--internal-ip',
+        '--zone={}'.format(client_worker._zone),
+        ' pytorchtpudistrunner@{}'.format(client_worker._hostname),
+        ' --command="{}"'.format(' '.join(cleanup_cmd)),
+    ]
+    proc = subprocess.Popen(' '.join(docker_rm_cmd), stdout=subprocess.PIPE,
```

Comment:
No need to concat, `Popen()` takes a list as well.

Reply:
`shell=True` as `Popen()` arg?

Reply:
The issue is that you are using `&&` which is a shell thing.
I think it is better to run the commands one-by-one, and not use `&&`.
Also, the run&wait_for_logs seems to be repeating.
Maybe it is better to have an API.

Reply:
Can I see the code?

Reply:
But that is not the complete code right?
You do, in your code, a sequence of:

1. rm script
2. docker rm
3. pkill

What I am saying to to have one API which wraps the repeated code (running and streaming logs), and run the three commands independently.